Dynamic heap for ESP Wi-Fi

marekmatej opened the pull request:
The Zephyr's system heap is allocated statically and it is not possible to add the memory areas which are left unoccupied after all static allocations and bootstrapping.

This PR adds the possibility to re-claim those areas dynamically.
- Added Kconfig `CONFIG_ESP_RUNTIME_HEAP` (enabled by default)
- Utilize memory left after all static allocations and use it as Wi-Fi heap.
- Added choice Kconfig `CONFIG_ESP_WIFI_HEAP_*` to select the heap that WiFi adapter is using. Value `CONFIG_ESP_WIFI_HEAP_RUNTIME` is the default option.

All the additional system heap increments can be freed and used for other purposes. The impact on DRAM usage for the most commonly used ESP32 boards is summarized here. (The `samples/net/wifi` was used).

#### `ESP32-S3_DevkitC/esp32/procpu`
**Without** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:       46796 B       329 KB     13.89%
     dram0_0_seg:      163368 B       313 KB     50.97%
```

**With** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:       46852 B       329 KB     13.91%
     dram0_0_seg:      116336 B       313 KB     36.30%
```

#### `ESP32_Devkitc_wroom/esp32/procpu`

**Without** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:       64768 B       235 KB     26.91%
     dram0_0_seg:      124128 B       192 KB     63.13%
```

**With** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:       65024 B       224 KB     28.35%
     dram0_0_seg:       81408 B       192 KB     41.41%
```

#### `ESP32C3_DevkitC`
     
**without** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:      117264 B     301568 B     38.88%
     dram0_0_seg:      240320 B     301568 B     79.69%
```
     
**With** the dynamic ESP heap:
```
Memory region         Used Size  Region Size  %age Used
     iram0_0_seg:      117328 B     301568 B     38.91%
     dram0_0_seg:      197392 B     301568 B     65.46%
```

marekmatej committed:
soc: espressif: Introduce runtime heap mempool

Add the `CONFIG_ESP_RUNTIME_HEAP` kconfig.
This allows the memory pool to be created starting
at `z_mapped_end` ending at `_heap_sentry`.

Added choice symbol ESP_WIFI_HEAP_* to select which
heap to use in the ESP WiFi adapter module.

Add file heap.c with code to initialize the runtime heap.
Size of the pool is checked during the runtime.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
soc: esp32c3: ESP WiFi heap

Provide symbols for the creation of dynamic memory pool.
Fix static allocations size check.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
manifest: update hal_espressif

Get latest necessary changes to support the following changes.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
boards: espressif: Update system heap increment

Remove the system heap increment based on used parts.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
samples: wifi: ESP32-S2 sample configuration

Remove the system heap override config from the board sample config.
Remove the memory saving restrictions.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
soc: esp32: ESP WiFi heap

Provide symbols for the creation of dynamic memory pool.
Fix static allocation size check.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
soc: esp32s2: ESP WiFi heap

Provide symbols for the creation of dynamic memory pool.
Fix the loader ROM buffers start address.
Fix static allocation size check.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
soc: esp32s3: ESP WiFi heap

Provide symbols for the creation of dynamic memory pool.
Fix static allocations size check.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

marekmatej committed:
soc: esp32c2: ESP WiFi heap

Provide symbols for the creation of dynamic memory pool.
Update the ROM-code SRAM usage according the IDF main.
Fix static allocations size check.
Increase iram_seg memory size for MCUboot.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

sylvioalves commented on soc/espressif/esp32s3/default.ld:
```diff
@@ -952,8 +945,8 @@ SECTIONS
 
-ASSERT(((_heap_start - ORIGIN(dram0_0_seg)) <= LENGTH(dram0_0_seg)),
-          "DRAM segment data does not fit.")
+//ASSERT(((_heap_start - ORIGIN(dram0_0_seg)) <= LENGTH(dram0_0_seg)),
```
Should this be removed?

sylvioalves commented:
Overall LGTM. I have a question though: What if `CONFIG_ESP32_WIFI_HEAP_MEM_POOL=n`? Of course I understand this situation would not be recommended but it is an option. I think this would block Wi-Fi from working properly. Can you confirm?

marekmatej commented:
> Overall LGTM. I have a question though: What if `CONFIG_ESP32_WIFI_HEAP_MEM_POOL=n`? Of course I understand this situation would not be recommended but it is an option. I think this would block Wi-Fi from working properly. Can you confirm?

In case `CONFIG_ESP32_WIFI_HEAP_MEM_POOL=n`, the default system heap will be used, and it is up to the system developer to have enough resources to accommodate WiFi.

However, in subsequent PR I would like to provide more options for WiFi heap, so it won't be limited by the dynamic heap size or usage.

marekmatej committed:
soc: esp32c6: Add runtime heap symbols

Update the linker scripts to provide necessary symbols.
Fix static allocation size check.

Signed-off-by: Marek Matej <marek.matej@espressif.com>

wmrsouza reviewed (changes requested):
Please, include ESP32-S3-DevKitC

marekmatej commented:
> Please, include ESP32-S3-DevKitC

Thx, I missed that board. Updated! Please revisit.

sylvioalves commented on soc/espressif/common/Kconfig:
```diff
@@ -24,6 +24,15 @@ config ESP_SIMPLE_BOOT
+	  ending by a last memory location that can be safely accesed (depending on a boot mode).
+	  This is a memory pool used in runtime to create a new heap memory.
+
+
```
Extra line here
